lib/src/kvs/clock.rs

```diff
@@ -0,0 +1,86 @@
+use crate::dbs::node::Timestamp;
+use crate::sql;
+use sql::Duration;
+use std::sync::Arc;
+use std::time::{SystemTime, UNIX_EPOCH};
+use tokio::sync::Mutex;
+
+// Traits cannot have async and we need sized structs for Clone + Send + Sync
+pub enum SizedClock {
+	Fake(FakeClock),
+	Inc(IncFakeClock),
+	System(SystemClock),
+}
+
+/// FakeClock is a clock that is fully controlled externally.
+/// Use this clock for when you are testing timestamps.
+#[derive(Clone)]
+pub struct FakeClock {
+	now: Timestamp,
+}
+
+impl FakeClock {
+	pub fn new(now: Timestamp) -> Self {
+		FakeClock {
+			now,
+		}
+	}
+
+	pub fn now(&self) -> Timestamp {
+		self.now.clone()
+	}
+
+	pub fn set(&mut self, timestamp: Timestamp) {
+		self.now = timestamp;
+	}
+}
+
+/// IncFakeClock increments a local clock every time the clock is accessed, similar to a real clock.
+/// This is useful when you need unique and partially deterministic timestamps for tests.
+/// Partially deterministic, because you do not have direct control over how many times a clock
+/// is accessed, and due to the nature of async - you neither have order guarantee.
+#[derive(Clone)]
+pub struct IncFakeClock {
+	now: Arc<Mutex<Timestamp>>,
```

Comment:
I see many instances of the following snippet in the tests:

```
	let clock = SizedClock::Fake(FakeClock::new(old_time.clone()));
	let clock = Arc::new(RwLock::new(clock));
```

If we wrap the entire SizedClock into Arc<Mutex> or Arc<RwLock> when we need mutability of the clock, why do we need `now: Arc<Mutex<Timestamp>>`?

Can it be `now: Timestamp` as you've made `FakeClock` so, or am I missing something that justifies making the now field wrapped?!

Reply:
Ah you are right! I was thinking that we want `IncFakeClock` to be thread safe, and tests to be thread safe. But if the tests are thread safe, then we can lessen the fake clock 👍

Reply:
So, I have an answer.
The clock must be cloneable, so we need Arc.
And it must be mutable, so we need a guard.
I have changed mutex to RwLock though. I will also check outside references and simplify since we don't need those presumably?

Reply:
Thanks for your reply!
Not sure if I understand it correctly, but if you are saying IncFakeClock needs to be cloneable-- can we perhaps make `Timestamp` cloneable so that we don't need `Arc<Mutex<Timestamp>` here?

Reply:
I have changed FakeClock to be `Arc<RwLock<Timestamp>>` but everything surrounding it I have made not rely on locks 🙌  Thanks